x/virtualgroup/keeper/keeper.go

```diff
@@ -481,6 +490,76 @@ func (k Keeper) StorageProviderExitable(ctx sdk.Context, spID uint32) error {
 	return nil
 }
 
+func (k Keeper) GetOrCreateGVGFamilyStatisticsWithinSP(ctx sdk.Context, spID uint32) *types.GVGFamilyStatisticsWithinSP {
+	store := ctx.KVStore(k.storeKey)
+
+	ret := &types.GVGFamilyStatisticsWithinSP{
+		StorageProviderId: spID,
+	}
+	bz := store.Get(types.GetGVGFamilyStatisticsWithinSPKey(spID))
+	if bz == nil {
+		return ret
+	}
+
+	k.cdc.MustUnmarshal(bz, ret)
+	return ret
+}
+
+func (k Keeper) DeleteGVGFamilyWithinSP(ctx sdk.Context, spID uint32, familyID uint32) {
+	gvgFamilyStatistics := k.MustGetGVGFamilyStatisticsWithinSP(ctx, spID)
+	index := -1
+	for i, id := range gvgFamilyStatistics.GlobalVirtualGroupFamilyIds {
+		if id == familyID {
+			index = i
+			break
+		}
+	}
+	if index != -1 {
+		gvgFamilyStatistics.GlobalVirtualGroupFamilyIds = append(gvgFamilyStatistics.GlobalVirtualGroupFamilyIds[:index], gvgFamilyStatistics.GlobalVirtualGroupFamilyIds[index+1:]...)
+		k.SetGVGFamilyStatisticsWithinSP(ctx, gvgFamilyStatistics)
+	}
+}
+
+func (k Keeper) GetGVGFamilyStatisticsWithinSP(ctx sdk.Context, spID uint32) (*types.GVGFamilyStatisticsWithinSP, bool) {
+	store := ctx.KVStore(k.storeKey)
+
+	bz := store.Get(types.GetGVGFamilyStatisticsWithinSPKey(spID))
+	if bz == nil {
+		return nil, false
+	}
+
+	var ret types.GVGFamilyStatisticsWithinSP
+	k.cdc.MustUnmarshal(bz, &ret)
+	return &ret, true
+}
+
+func (k Keeper) MustGetGVGFamilyStatisticsWithinSP(ctx sdk.Context, spID uint32) *types.GVGFamilyStatisticsWithinSP {
+	store := ctx.KVStore(k.storeKey)
+
+	bz := store.Get(types.GetGVGFamilyStatisticsWithinSPKey(spID))
+	if bz == nil {
+		panic("must get vgf within sp")
+	}
+
+	var ret types.GVGFamilyStatisticsWithinSP
+	k.cdc.MustUnmarshal(bz, &ret)
+	return &ret
+}
+
+func (k Keeper) SetGVGFamilyStatisticsWithinSP(ctx sdk.Context, vgfStatisticsWithinSP *types.GVGFamilyStatisticsWithinSP) {
+	store := ctx.KVStore(k.storeKey)
+
+	bz := k.cdc.MustMarshal(vgfStatisticsWithinSP)
+
+	store.Set(types.GetGVGFamilyStatisticsWithinSPKey(vgfStatisticsWithinSP.StorageProviderId), bz)
+}
+
+func (k Keeper) BatchSetGVGFamilyStatisticsWithinSP(ctx sdk.Context, vgfStatisticsWithinSP []*types.GVGFamilyStatisticsWithinSP) {
```

Comment:
need to populate data into store as new key is used.

Reply:
update the logic to set all GVGFamilyStatisticsWithinSP when executing the hardfork